s3/libirods_s3.cpp

```diff
@@ -494,121 +506,126 @@ irods::error s3ReadAuthInfo(
-            retry_count_str );
-        if( ret.ok() ) {
-            try {
-                g_retry_count = boost::lexical_cast<int>( retry_count_str );
-            } catch ( const boost::bad_lexical_cast& ) {
-                rodsLog(
-                    LOG_ERROR,
-                    "failed to cast retry count [%s] to an int",
-                    retry_count_str.c_str() );
-            }
+    // First, parse the default hostname (if present) into a list of
+    // hostnames separated on the definition line by commas (,)
+    std::string hostname_list;
+    irods::error ret;
+    ret = _prop_map.get< std::string >(
+        s3_default_hostname,
+        hostname_list );
+    if( !ret.ok() ) {
+        // ok to fail
+        hostname_vector.push_back(S3_DEFAULT_HOSTNAME); // Default to Amazon
+    } else {
+        std::stringstream ss(hostname_list);
+        std::string item;
+        while (std::getline(ss, item, ',')) {
+            hostname_vector.push_back(item);
         }
+        // Because each resource operation is a new instance, randomize the starting
+        // hostname offset so we don't always hit the first in the list between different
+        // operations.
+        srand(time(NULL));
+        hostname_index = rand() % hostname_vector.size();
+    }
 
-        std::string wait_time_str;
-        ret = _prop_map.get< std::string >(
-            s3_wait_time_sec,
-            wait_time_str );
-        if( ret.ok() ) {
-            try {
-                g_retry_wait = boost::lexical_cast<int>( wait_time_str );
-            } catch ( const boost::bad_lexical_cast& ) {
-                rodsLog(
-                    LOG_ERROR,
-                    "failed to cast wait time [%s] to an int",
-                    wait_time_str.c_str() );
-            }
+    _prop_map.set<std::vector<std::string> >(s3_default_hostname_vector, hostname_vector);
+    _prop_map.set<size_t>(s3_hostname_index, hostname_index);
+
+    size_t retry_count = 10;
+    std::string retry_count_str;
+    ret = _prop_map.get< std::string >(
+        s3_retry_count,
+        retry_count_str );
+    if( ret.ok() ) {
+        try {
+            retry_count = boost::lexical_cast<int>( retry_count_str );
+        } catch ( const boost::bad_lexical_cast& ) {
+            rodsLog(
+                LOG_ERROR,
+                "failed to cast retry count [%s] to an int",
+                retry_count_str.c_str() );
         }
+    }
+    _prop_map.set<size_t>(s3_retry_count_size_t, retry_count);
 
-        size_t ctr = 0;
-        while( ctr < g_retry_count ) {
-            int status = 0;
-            int flags = S3_INIT_ALL;
-            S3SignatureVersion signature_version = s3GetSignatureVersion(_prop_map);
+    size_t wait_time = 1;
+    std::string wait_time_str;
+    ret = _prop_map.get< std::string >(
+        s3_wait_time_sec,
+        wait_time_str );
+    if( ret.ok() ) {
+        try {
+            wait_time = boost::lexical_cast<int>( wait_time_str );
```

Comment:
`wait_time` is a `size_t` but we are lexical casting to `int` (and then implicitly back to `size_t`). What happens when the value is negative? Should we use a `decltype(wait_time)` or simply a `size_t` in the template parameter for consistency?

Reply:
Not changed code (other than indention) but will fix.